src/workerd/api/streams/writable.c++

```diff
@@ -369,6 +369,173 @@ private:
+      KJ_IF_SOME(obj, weakRef->tryGet()) {
+        // Stream is still alive, revoke it.
+        if (!obj.canceler.isEmpty()) {
+          obj.canceler.cancel(cancellationException());
+        }
+      }
+    }));
+  }
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    if (size == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run([this, buffer, size](Worker::Lock& lock) mutable {
+      jsg::Lock& js = lock;
+      // Sadly, we have to allocate and copy here. Our received buffer is only guaranteed
+      // to live until the returned promise is resolved, but the application code may
+      // hold onto the ArrayBuffer for longer. We need to make sure that the backing store
+      // for the ArrayBuffer remains valid.
+      auto source = KJ_ASSERT_NONNULL(jsg::BufferSource::tryAlloc(js, size));
+      auto ptr = source.asArrayPtr();
+      memcpy(ptr.begin(), buffer, size);
+      return context.awaitJs(js, getInner().write(js, source.getHandle(js)));
+    }));
+  }
+
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+    auto amount = 0;
+    for (auto& piece : pieces) {
+      amount += piece.size();
+    }
+    if (amount == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run([this, amount, pieces](Worker::Lock& lock) mutable {
+      jsg::Lock& js = lock;
+
+      // Sadly, we have to allocate and copy here. Our received set of buffers are only
+      // guaranteed to live until the returned promise is resolved, but the application code
+      // may hold onto the ArrayBuffer for longer. We need to make sure that the backing store
+      // for the ArrayBuffer remains valid.
+      auto source = KJ_ASSERT_NONNULL(jsg::BufferSource::tryAlloc(js, amount));
+      auto ptr = source.asArrayPtr();
+      for (auto& piece : pieces) {
+        if (piece.size() == 0) continue;
+        KJ_ASSERT(piece.size() <= ptr.size());
+        memcpy(ptr.begin(), piece.begin(), piece.size());
+        ptr = ptr.slice(piece.size());
+      }
+
+      return context.awaitJs(js, getInner().write(js, source.getHandle(js)));
+    }));
+  }
+
+  // TODO(perf): We can't properly implement tryPumpFrom(), which means that Cap'n Proto will
+  //   be unable to perform path shortening if the underlying stream turns out to be another capnp
+  //   stream. This isn't a huge deal, but might be nice to enable someday. It may require
+  //   significant refactoring of streams.
+
+  kj::Promise<void> whenWriteDisconnected() override {
+    // TODO(soon): We might be able to support this by following the writer.closed promise,
+    // which becomes resolved when the writer is used to close the stream, or rejects when
+    // the stream has errored. However, currently, we don't have an easy way to do this for
+    // a couple of reasons.
+    //
+    // 1. The Writer's getClosed() method returns a jsg::MemoizedIdentity<jsg::Promise<void>>.
```

Comment:
This is actually somewhat of an advantage since `jsg::Promise<T>` is single-use but a true V8 promise can have multiple waiters. The only problem is that we don't have a convenient JSG wrapper for a raw V8 promise -- that's actually something I found myself wanting a bunch while writing the RPC code. Essentially a `JsValue` variant for promises, probably called `JsPromise`?

Reply:
Sounds reasonable.

Reply:
For now, I'll leave this method unimplemented and will tackle this problem in a follow up PR.